autogen/coding/base.py

```diff
@@ -0,0 +1,94 @@
+from typing import Any, Dict, List, Protocol, Union, runtime_checkable
+
+from pydantic import BaseModel, Field
+
+from ..agentchat.agent import LLMAgent
+
+__all__ = ("CodeBlock", "CodeResult", "CodeExtractor", "CodeExecutor")
+
+
+class CodeBlock(BaseModel):
+    """A class that represents a code block."""
+
+    code: str = Field(description="The code to execute.")
+
+    language: str = Field(description="The language of the code.")
+
+
+class CodeResult(BaseModel):
+    """A class that represents the result of a code execution."""
+
+    exit_code: int = Field(description="The exit code of the code execution.")
+
+    output: str = Field(description="The output of the code execution.")
+
+
+class CodeExtractor(Protocol):
+    """A code extractor class that extracts code blocks from a message."""
+
+    def extract_code_blocks(self, message: Union[str, List[Dict[str, Any]], None]) -> List[CodeBlock]:
+        """Extract code blocks from a message.
+
+        Args:
+            message (str): The message to extract code blocks from.
+
+        Returns:
+            List[CodeBlock]: The extracted code blocks.
+        """
+        ...  # pragma: no cover
+
+
+@runtime_checkable
+class CodeExecutor(Protocol):
+    """A code executor class that executes code blocks and returns the result."""
+
+    class UserCapability(Protocol):
+        """An AgentCapability class that gives agent ability use this code executor."""
+
+        def add_to_agent(self, agent: LLMAgent) -> None:
```

Comment:
`user_proxy.code_executor.user_capability.add_to_agent(agent)`  😵‍💫

It is a nice-to-have feature, but the learning curve might be steep here...

We should have "beginner", "intermediate", and "advanced" AutoGen tutorial. This part should belong to the "advanced" part 😂

Reply:
Right. The concept of Agent Capability is used by Teachability and also in a separate PR #1513. We will see more of this. 

This line of code is likely going to be wrapped by some top-level function that binds user proxy and assistant agent.